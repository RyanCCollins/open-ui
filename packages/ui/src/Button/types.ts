import { Props } from './';

export { Props };
export type Size = 'xsmall' | 'small' | 'medium' | 'large' | 'xlarge';
