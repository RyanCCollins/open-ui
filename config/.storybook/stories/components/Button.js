import React from 'react';
import { storiesOf, action } from '@kadira/storybook';
import { Button } from 'ui';
import {
  withKnobs,
  boolean,
  color,
  select,
  text
} from '@kadira/storybook-addon-knobs';


const stories = storiesOf('Button', module);
stories.addDecorator(withKnobs);

stories
  .addWithInfo(
    'default Button',
    `
      Properties:

      \`size: xsmall|small|medium|large|xlarge \`
      
      Button size

      \` borderColor: {string}\`
      
      Border color for the button

      \`backgroundColor: {string} \`
      
      Background color for the button

      \`Usage:\` 
      ~~~js
        <Button 
          onClick={this.props.handleClick} 
        >
          Default Button
        </Button>
      ~~~
    `,
     () => (
      <Button
        onClick={action('clicked the default button')}
        borderColor={color('borderColor','#c05b4d')}
        backgroundColor={color('backgroundColor','#c05b4d')}
      >
        {text('Label','Default Button')}
      </Button>
     ),
     {
       inline: true,
       propTables: false,
     }
  )
  .addWithInfo( 
    'hero Button',
    `
      Properties:

      \`size: xsmall|small|medium|large|xlarge \`
      
      Button size

      \`isHero: true|false\`

      Whether the button is a Hero
      
      \` borderColor: {string}\`
      
      Border color for the button

      \`backgroundColor: {string} \`
      
      Background color for the button
      
      ~~~js
        <Button 
          isHero,
          onClick={this.props.handleClick} 
        >
          Hero Button
        </Button>
      ~~~
    `,
     () => (
      <Button
        onClick={action('clicked the hero button')}
        style={{ margin: 10 }}
        size="xlarge"
        isHero={boolean('isHero','true')}
        backgroundColor={color('backgroundColor','#c05b4d')}
        borderColor={color('borderColor','#732419')}
      >
        {text('Label','Hero Button')}
      </Button>
     ),
     {
       inline: true,
       propTables: false,
     }
  );
