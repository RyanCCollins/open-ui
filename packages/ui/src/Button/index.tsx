import * as React from 'react';
const { HeroButton, ButtonComponent} = require('./styles');
import { Size } from './types';

export interface Props extends React.Props<Button> {
  color?: string;
  backgroundColor?: string;
  onClick?: (x: any) => any;
  size?: Size;
  isHero?: boolean;
  borderColor?: string;
  style?: any;
}

class Button extends React.Component<Props, undefined> {
  public static defaultProps: Props = {
    color: '#fefefe',
    backgroundColor: '#c05b4d',
    borderColor: '#732419',
    size: 'medium',
    isHero: false,
    style: {},
  };
  public render() {
    if (this.props.isHero) {
      return (
        <HeroButton
          style={this.props.style}
          borderColor={this.props.borderColor}
          onClick={this.props.onClick}
          color={this.props.color}
          backgroundColor={this.props.backgroundColor}
          size={this.props.size}
        >
          {this.props.children}
        </HeroButton>
      );
    }
    return (
      <ButtonComponent
        onClick={this.props.onClick}
        color={this.props.color}
        backgroundColor={this.props.backgroundColor}
        size={this.props.size}
      >
        {this.props.children}
      </ButtonComponent>
    );
  }
}

export default Button;
