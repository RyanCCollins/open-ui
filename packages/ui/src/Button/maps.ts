export const sizeMap = {
  xsmall: 10,
  small: 12,
  medium: 16,
  large: 20,
  xlarge: 24,
};
