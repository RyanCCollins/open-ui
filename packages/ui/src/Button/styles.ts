const styled = require('styled-components').default;
import remStringFromPX from '../utils';
import { sizeMap } from './maps';

export const ButtonComponent = styled.button`
  text-align: center;
  background-color: ${(props) => props.backgroundColor};
  color: ${(props) => props.color};
  font-size: ${(props) => remStringFromPX(sizeMap[props.size])};
  border: none;
  color: white;
  padding: 15px 32px;
  text-align: center;
  text-decoration: none;
  display: inline-block;
  cursor: pointer;
`;

export const HeroButton = styled.button`
  background-color: ${(props) => props.backgroundColor};
  color: ${(props) => props.color};
  padding: 20px 20px;
  line-height: 16px;
  font-size: ${(props) => remStringFromPX(sizeMap[props.size])};
  transition: 150ms;
  display: inline-block;
  border-style: solid;
  border-width: 0 0 4px 0;
  border-color: ${(props) => props.borderColor};
  border-radius: 3px;
  text-decoration: none;
  box-shadow: none;
  cursor: pointer;
  &:active {
    background-color: ${(props) => props.borderColor};
  }
`;
