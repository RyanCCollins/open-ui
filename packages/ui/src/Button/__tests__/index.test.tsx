import { shallow, mount } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import * as React from 'react';
import Button from '../';
import { buttonProps, heroButtonProps } from './__mocks__/buttonMocks.mock';

describe('<Button /> default', () => {
  it('should render a plain button with default props', () => {
    const wrapper = shallow(
      <Button>
        Button Text
      </Button>,
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
  it('should render a plain button with other props', () => {
    const wrapper = shallow(
      <Button {...buttonProps}>
        Button Text
      </Button>,
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});

describe('<Button /> when a hero button', () => {
  it('should render a hero button with boolean true', () => {
    const wrapper = shallow(
      <Button isHero={true}>
        Button Text
      </Button>,
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
  it('should render a hero button with other props', () => {
    const wrapper = shallow(
      <Button {...heroButtonProps}>
        Button Text
      </Button>,
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});

describe('Plain button is not a hero button', () => {
  let button;

  beforeEach(() => {
    button = mount(
      <Button>
        Button Text
      </Button>,
    );
  });

  it('Button requires isHero prop', () => {
    expect(button.props().isHero).toBeDefined();
    expect(button.props().isHero).toEqual(false);
  });
});

describe('onClick event for plain button', () => {
  let mockFunction;
  let button;

  beforeEach(() => {
    mockFunction = jest.fn();
    button = mount(
      <Button onClick={mockFunction}>
        Button Text
      </Button>,
    );
  });

  it('Button requires onClick prop', () => {
    expect(button.props().onClick).toBeDefined();
  });

  it('Button renders button', () => {
    const renderedButton = button.find('button').first();
    expect(renderedButton).toBeDefined();
  });

  it('Button click calls mockFunction', () => {
    const renderedButton = button.find('button').first();
    renderedButton.simulate('click');
    expect(mockFunction).toBeCalled();
  });
});

describe('onClick event for hero button', () => {
  let mockFunction;
  let button;

  beforeEach(() => {
    mockFunction = jest.fn();
    button = mount(
      <Button isHero={true} onClick={mockFunction}>
        Button Text
      </Button>,
    );
  });

  it('Button requires onClick prop', () => {
    expect(button.props().onClick).toBeDefined();
  });

  it('Button renders button', () => {
    const renderedButton = button.find('button').first();
    expect(renderedButton).toBeDefined();
  });

  it('Button click calls mockFunction', () => {
    const renderedButton = button.find('button').first();
    renderedButton.simulate('click');
    expect(mockFunction).toBeCalled();
  });
});
