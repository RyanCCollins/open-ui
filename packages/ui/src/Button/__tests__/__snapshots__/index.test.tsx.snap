// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> default should render a plain button with default props 1`] = `
<styled.button
  backgroundColor="#c05b4d"
  color="#fefefe"
  size="medium"
>
  Button Text
</styled.button>
`;

exports[`<Button /> default should render a plain button with other props 1`] = `
<styled.button
  backgroundColor="#ffffff"
  color="blue"
  size="xlarge"
>
  Button Text
</styled.button>
`;

exports[`<Button /> when a hero button should render a hero button with boolean true 1`] = `
<styled.button
  backgroundColor="#c05b4d"
  borderColor="#732419"
  color="#fefefe"
  size="medium"
  style={Object {}}
>
  Button Text
</styled.button>
`;

exports[`<Button /> when a hero button should render a hero button with other props 1`] = `
<styled.button
  backgroundColor="#ffffff"
  borderColor="#000000"
  color="blue"
  size="large"
  style="text-align: center;"
>
  Button Text
</styled.button>
`;
